### ISSUE TITLE
Crash

### ISSUE THREAD

**Issue by markus2330** [linux-pam/linux-pam#275]
Forwarded from https://github.com/i3/i3lock/issues/292

## I'm submitting a…

<pre>
[X] Bug
[ ] Feature Request
[ ] Other (Please describe in detail)
</pre>

## Current Behavior

I typed in the password in i3lock (I think the password was correct), pressed Enter (I am not sure if they key stroke was recognized) but still the green circle was shown. Then I pressed enter (again?) and the program crashed:

```
Core was generated by `i3lock -c 000000 -n'.
Program terminated with signal SIGABRT, Aborted.
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007f06a0055535 in __GI_abort () at abort.c:79
#2  0x00007f06a00ac508 in __libc_message (action=action@entry=do_abort, fmt=fmt@entry=0x7f06a01b728d "%s\n") at ../sysdeps/posix/libc_fatal.c:181
#3  0x00007f06a00b2c1a in malloc_printerr (str=str@entry=0x7f06a01b8f58 "free(): double free detected in tcache 2") at malloc.c:5341
#4  0x00007f06a00b46fd in _int_free (av=0x7f06a01eec40 <main_arena>, p=0x5606dac53220, have_lock=<optimized out>) at malloc.c:4193
#5  0x00007f06a039b6e4 in pam_set_item () from /lib/x86_64-linux-gnu/libpam.so.0
#6  0x00007f06a039c31f in ?? () from /lib/x86_64-linux-gnu/libpam.so.0
#7  0x00007f06a0397659 in pam_authenticate () from /lib/x86_64-linux-gnu/libpam.so.0
#8  0x00005606da0b2191 in ?? ()
#9  0x00005606da0b2a3b in ?? ()
#10 0x00007f06a037bf43 in ev_invoke_pending () from /usr/lib/x86_64-linux-gnu/libev.so.4
#11 0x00007f06a037f7e5 in ev_run () from /usr/lib/x86_64-linux-gnu/libev.so.4
#12 0x00005606da0b1446 in ?? ()
#13 0x00007f06a005709b in __libc_start_main (main=0x5606da0b0b90, argc=4, argv=0x7ffc0a7ea318, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7ffc0a7ea308) at ../csu/libc-start.c:308
#14 0x00005606da0b187a in ?? ()
```

## Expected Behavior

No crash.

## Reproduction Instructions

I could not reproduce it.

## Environment
Output of `i3lock --version`:
<pre>
i3lock: version 2.11.1 © 2010 Michael Stapelberg
</pre>

libpam0g version on Debian Stretch 10.6:
```
1.3.1-5
```

**Comment by t8m** [comment]
Is it a reproducible crash, or did that happen only once?

**Comment by t8m** [comment]
It is very unlikely the bug is with libpam as we do not have reports with similar backtraces from other applications using libpam.

**Comment by markus2330** [comment]
Unfortunately I am unable to reproduce it, it only happened once after long use of the tool. I also see the problem in i3lock, I reported the problem here only because I was requested to do so.

To make it easier to find such a problem, it would, however, maybe make sense to have assertions in `pam_end` and similar to find misuse of the PAM API more easily.

**Comment by markus2330** [comment]
For me it is plausible that I simply ran into a known race condition in i3lock https://github.com/i3/i3lock/issues/254, which was hopefully fixed in https://github.com/i3/i3lock/commit/656fa39f82107441c978a9a7397ebbea58e06a0d (I am not sure sure if they use `pam_end` correctly, however, but this is another issue). I think we can close this issue.
